contracts/smart-wallet/utils/AccountCore.sol

```diff
@@ -70,8 +75,46 @@ contract AccountCore is Initializable, Multicall, BaseAccount, ERC1271 {
     }
 
     /// @notice Returns whether a signer is authorized to perform transactions using the wallet.
-    function isValidSigner(address _signer) public view virtual returns (bool) {
-        return _hasRole(SIGNER_ROLE, _signer) || _hasRole(DEFAULT_ADMIN_ROLE, _signer);
+    function isValidSigner(address _signer, UserOperation calldata _userOp) public view virtual returns (bool) {
+        AccountPermissionsStorage.Data storage data = AccountPermissionsStorage.accountPermissionsStorage();
+
+        // First, check if the signer is an admin.
+        if (data.isAdmin[_signer]) {
+            return true;
+        } else {
+            // If not an admin, check restrictions for the role held by the signer.
+            bytes32 role = data.roleOfAccount[_signer];
+            Role memory restrictions = data.roleRestriction[role];
+
+            // Check if the role is active. If the signer has no role, this condition will revert because both start and end timestamps are `0`.
+            require(
+                restrictions.startTimestamp <= block.timestamp && block.timestamp < restrictions.endTimestamp,
+                "Account: role not active."
+            );
+
+            // Extract the function signature from the userOp calldata and check whether the signer is attempting to call `execute` or `executeBatch`.
+            bytes4 sig = getFunctionSignature(_userOp.callData);
+
+            if (sig == Account.execute.selector) {
+                // Extract the `target` and `value` arguments from the calldata for `execute`.
+                (address target, uint256 value) = decodeExecuteCalldata(_userOp.callData);
+
+                // Check if the value is within the allowed range and if the target is approved.
+                require(restrictions.maxValuePerTransaction >= value, "Account: value too high.");
+                require(data.approvedTargets[role].contains(target), "Account: target not approved.");
+            } else if (sig == Account.executeBatch.selector) {
+                // Extract the `target` and `value` array arguments from the calldata for `executeBatch`.
+                (address[] memory targets, uint256[] memory values) = decodeExecuteBatchCalldata(_userOp.callData);
+
+                // For each target+value pair, check if the value is within the allowed range and if the target is approved.
+                for (uint256 i = 0; i < targets.length; i++) {
+                    require(data.approvedTargets[role].contains(targets[i]), "Account: target not approved.");
+                    require(restrictions.maxValuePerTransaction >= values[i], "Account: value too high.");
+                }
+            }
```

Comment:
shouldn't we revert if the function called is neither execute/executeBatch here?